### PR TITLE
[AP-1028] Refactor to use confluent-kafka

### PR DESCRIPTION
**Pull request by koszti**
## Problem

`pipelinewise-tap-kafka` is currently using [kafka-python](https://github.com/dpkp/kafka-python) library. We'd like to use [confluent-kafka-python](https://github.com/confluentinc/confluent-kafka-python) because of the better support from confluent. The two libraries are similar to use but requires refactoring of the current implementation

## Proposed changes

Refactor code to use `confluent-kafka` instead of `kafka-python` lib.

## Types of changes

What types of changes does your code introduce to PipelineWise?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [x] Description above provides context of the change
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Unit tests for changes (not needed for documentation changes)
- [x] CI checks pass with my changes
- [x] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [ ] Commit message/PR title starts with `[AP-NNNN]` (if applicable. AP-NNNN = JIRA ID)
- [ ] Branch name starts with `AP-NNN` (if applicable. AP-NNN = JIRA ID)
- [ ] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -66,7 +66,6 @@ Full list of options in `config.json`:
 | consumer_timeout_ms                 | Integer |            | (Default: 10000) KafkaConsumer setting. Number of milliseconds to block during message iteration before raising StopIteration            |
 | session_timeout_ms                  | Integer |            | (Default: 30000) KafkaConsumer setting. The timeout used to detect failures when using Kafka’s group management facilities. |                                      |
 | heartbeat_interval_ms               | Integer |            | (Default: 10000) KafkaConsumer setting. The expected time in milliseconds between heartbeats to the consumer coordinator when using Kafka’s group management facilities. |
-| max_poll_records                    | Integer |            | (Default: 500) KafkaConsumer setting. The maximum number of records returned in a single call to poll(). |
```
Comment (Samira-El) [inline README.md:69]: Why delete this? is it being deprecated?
There is a tap in prod that is using it.

File: tap_kafka/__init__.py [hunk 2]
```diff
@@ -40,20 +42,32 @@ def dump_catalog(all_streams):
...
+        'bootstrap.servers': config['bootstrap_servers'],
+        'group.id': config['group_id'],
+        'auto.offset.reset': 'earliest',
+    })
+
+    try:
+        topic = config['topic']
+
+        LOGGER.info(f"Discovering {topic} topic...")
+        cluster_md = consumer.list_topics(topic=topic, timeout=config['session_timeout_ms'] / 1000)
```
Comment (Samira-El) [inline tap_kafka/__init__.py:55]: this is making `session_timeout_ms` required when it's not according to the README.
There is a fallback default value to use `DEFAULT_SESSION_TIMEOUT_MS`

File: README.md (hunk 1 above)
Comment (koszti) [reply README.md:69]: This option is not supported by the by the underlying librdkafka C library. The full range of configuration properties please consult librdkafka’s documentation: https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md

More info in the [confluent-kafka documentation](https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#pythonclient-configuration)

File: tap_kafka/__init__.py (hunk 2 above)
Comment (koszti) [reply tap_kafka/__init__.py:55]: The `config` dict here is not the original `config.json` but a modified one where all the default values are replaced. This parameter should always exist at this point.

The original config modified by the [generate_config](https://github.com/transferwise/pipelinewise-tap-kafka/blob/2501966faf7a2afd8ecbeb8b3a875cd0a7ac7498/tap_kafka/__init__.py#L77) function and the output is passed to the [do_discover](https://github.com/transferwise/pipelinewise-tap-kafka/blob/2501966faf7a2afd8ecbeb8b3a875cd0a7ac7498/tap_kafka/__init__.py#L108) function

Comment (Samira-El) [reply tap_kafka/__init__.py:55]: aha! my bad!

File: tap_kafka/sync.py [hunk 3]
```diff
@@ -69,40 +69,40 @@ def init_local_store(kafka_config):
...
+        "message_offset": message.offset(),
+        "message_partition": message.partition()
     }
 
     # Add primary keys to the record message
     for key in primary_keys:
         pk_selector = primary_keys[key]
         try:
-            record[key] = dpath.util.get(message.value, pk_selector)
+            record[key] = dpath.util.get(message.value(), pk_selector)
```
Comment (Samira-El) [inline tap_kafka/sync.py:105]: I noticed that the code continues executing if PK is not found in the message which i think is not the right behavior, it also makes the target fail.

I know this isn't related to this PR, so perhaps we could have a ticket to update/discuss this behavior.

Comment (koszti) [reply tap_kafka/sync.py:105]: Thanks, I agree, we need to agree what's the expected behaviour and address separately out of this PR.

File: README.md (hunk 1 above)
Comment (Samira-El) [reply README.md:69]: Okay.

Can you remove the `DEFAULT_MAX_POLL_RECORDS` constant then?

**Pull request merged**